### PR TITLE
C++: IR construction for `VacuousDestructorCall`s

### PR DESCRIPTION
**Pull request by MathiasVP**
This reduces the number of `useNotDominatedByDefinition` inconsistencies on ChakraCore from ~13K to ~11.5K, and I suspect it'll have a big impact on quite a lot of databases.

This PR will conflict ever so slightly with https://github.com/github/codeql/pull/15506, so I suggest we first merge https://github.com/github/codeql/pull/15506 and then fix up this PR afterwards.

### CODE REVIEW

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected [hunk 1]
```diff
@@ -7,7 +7,6 @@ missingOperandType
 duplicateChiOperand
 sideEffectWithoutPrimary
 instructionWithoutSuccessor
-| VacuousDestructorCall.cpp:2:29:2:29 | InitializeIndirection: y | Instruction 'InitializeIndirection: y' has no successors in function '$@'. | VacuousDestructorCall.cpp:2:6:2:6 | void CallDestructor<int>(int, int*) | void CallDestructor<int>(int, int*) |
```
Comment (jketema) [inline cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected:10]: We could have known :tableflip:

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected [hunk 2]
```diff
@@ -7,7 +7,6 @@ missingOperandType
...
 | ms_try_mix.cpp:35:13:35:19 | ThrowValue: throw ... | Instruction 'ThrowValue: throw ...' has no successors in function '$@'. | ms_try_mix.cpp:29:6:29:19 | void ms_finally_mix(int) | void ms_finally_mix(int) |
 | ms_try_mix.cpp:53:5:53:11 | ThrowValue: throw ... | Instruction 'ThrowValue: throw ...' has no successors in function '$@'. | ms_try_mix.cpp:49:6:49:28 | void ms_empty_finally_at_end() | void ms_empty_finally_at_end() |
 | stmt_expr.cpp:27:5:27:15 | Store: ... = ... | Instruction 'Store: ... = ...' has no successors in function '$@'. | stmt_expr.cpp:21:13:21:13 | void stmtexpr::g(int) | void stmtexpr::g(int) |
```
Comment (jketema) [inline cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected:12]: I wonder if we should look at this one too to see what is going on?

**Review by jketema — APPROVED** [review]
LGTM is DCA is happy.

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected:10]: Indeed 😂

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected:12]: Good idea! Looking at the file it suggests it has something to do with computed goto labels: https://github.com/github/codeql/blob/3ab6f222d0a405ec7ccc36631d4ecd67cd191744/cpp/ql/test/library-tests/syntax-zoo/stmt_expr.cpp#L27

Comment (jketema) [reply cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected:12]: That's pretty esoteric, so probably shouldn't take priority.

**Comment by MathiasVP** [comment]
The IR inconsistency reductions look really good 🎉

Screenshot 2024-02-27 at 17 30 13

**Review by geoffw0 — APPROVED** [review]
Re-approving.

**Pull request merged**